en/debug/system_wide_replay.md

```diff
@@ -1,93 +1,113 @@
 # System-wide Replay
 
-Based on ORB messages, it's possible to record and replay arbitrary parts of the system.
+It is possible to record and replay arbitrary parts of the system based on ORB messages.
 
 Replay is useful to test the effect of different parameter values based on real data, compare different estimators, etc.
 
 ## Prerequisites
 
-The first thing that needs to be done is to identify the module or modules that should be replayed.
+The first step is to identify the module or modules that should be replayed.
 Then, identify all the inputs to these modules, i.e. subscribed ORB topics.
 For system-wide replay, this consists of all hardware input: sensors, RC input, MAVLink commands and file system.
 
 All identified topics need to be logged at full rate (see [logging](../dev_log/logging.md)).
 For `ekf2` this is already the case with the default set of logged topics.
 
 It is important that all replayed topics contain only a single absolute timestamp, which is the automatically generated field `timestamp`.
-Should there be more timestamps, then they must be relative with respect to the main timestamp.
+Should there be more timestamps, they must be relative to the main timestamp.
 For an example, see [SensorCombined.msg](https://github.com/PX4/PX4-Autopilot/blob/main/msg/SensorCombined.msg).
 Reasons for this are given below.
 
 
 ## Usage
 
-- First, choose the file to replay, and build the target (from within the PX4-Autopilot directory):
+- First, choose the file to replay and build the target (from within the PX4-Autopilot directory):
   
   ```sh
   export replay=<absolute_path_to_log_file.ulg>
   make px4_sitl_default
   ```
-  This will create the output in a separate build directory `build/px4_sitl_default_replay` (so that the parameters don't interfere with normal builds).
-  It's possible to choose any posix SITL build target for replay, the build system knows through the `replay` environment variable that it's in replay mode.
-- Add ORB publisher rules file in `build/px4_sitl_default_replay/rootfs/orb_publisher.rules`.
-  This file defines which module is allowed to publish which messages.
+  This will create the build/make output in a separate build directory `build/px4_sitl_default_replay` (so that the parameters don't interfere with normal builds).
+  It's possible to choose any posix SITL build target for replay, since the build system knows through the `replay` environment variable that it's in replay mode.
+- Add ORB publisher rules in the file `build/px4_sitl_default_replay/rootfs/orb_publisher.rules`.
+  This file defines the modules that are allowed to publish particular messages.
   It has the following format:
+  
   ```
   restrict_topics: <topic1>, <topic2>, ..., <topicN>
   module: <module>
   ignore_others: <true/false>
   ```
```

Comment:
1. Can you only define one module? I.e what if you want to define several modules and restrictions?

2. My understanding is that these publisher rules are optional - you could just play the log file your recorded. You add publisher rules because (from below) "the modules that usually publish these topics don't need to be disabled for the replay." . I'm not sure I get this - why do you want to disable publications from one place and not the other?

Reply:
1. That is something I am not actually sure of - I have not tried it.

2. The rules probably are optional. That said, in the case of the EKF2 replay at least, the `replay` module republishes certain topics directly from the log. In this case, we want the module that originally published them to _not_ publish them such that there is only one source of those messages. These other modules can keep running, but if they publish on those topics we ignore it.

Reply:
Ah. That makes sense. Looking forward to an update, when and if you are able.